src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fs

```diff
@@ -377,13 +387,25 @@ type InferenceContext(symbolTracker: SymbolTracker) =
         | Callable (callable, input, output) ->
             let callable = context.Resolve callable
 
-            match callable.Resolution with
-            | QsTypeKind.Operation _ ->
-                let operationType = QsTypeKind.Operation((input, output), CallableInformation.NoInformation)
-                context.ConstrainImpl(callable <. ResolvedType.New operationType)
-            | QsTypeKind.Function _ ->
-                context.ConstrainImpl(callable <. ResolvedType.New(QsTypeKind.Function(input, output)))
-            | _ ->
+            let requiredCallable =
+                match callable.Resolution with
+                | QsTypeKind.Operation _ ->
+                    QsTypeKind.Operation((input, output), CallableInformation.NoInformation) |> ResolvedType.New |> Some
+                | QsTypeKind.Function _ -> QsTypeKind.Function(input, output) |> ResolvedType.New |> Some
+                | _ -> None
+
+            // The callable constraint needs special handling for diagnostics:
+            //
+            // - If the source is here, it is assumed to be a call expression. The callable is the "expected" type. A
+            //   type mismatch is the argument's fault.
+            // - If the source is somewhere else, the callable is being provided as a value to a higher-order function.
```

Comment:
Does this logic (in particular the second bullet) also apply to lambdas that may be declared in the same function?

Reply:
This special case was added for this test case involving a lambda:

```qsharp
let g = op => op();
g(X);
  ~
```

because `op` has a deferred callable constraint.